project/src/main/webapp/document.jsp

```diff
@@ -177,6 +187,173 @@
+        var markerList = [];
+        var widgetsFound = 0;
+
+        // Identify start and end points of comments and associate appropriate data
+        for(var lineIndex = 0; lineIndex < codeMirror.lineCount(); lineIndex++) {
+          var line = codeMirror.getLine(lineIndex);
+          for(var charIndex = 0; charIndex < line.length; charIndex++) {
+            if(line.charCodeAt(charIndex) > 255) {
+              // Grab the data inside the entity that was created
+              var widget = widgetElements[widgetsFound].getElementsByTagName("link")[0];
+              markerList.push({line: lineIndex, ch: charIndex, id: widget.getAttribute("data-id"), pos: widget.getAttribute("data-pos")});
+              widgetsFound++;
+            }
+          }
+        }
+
+        // Create comment stying then remove special characters
+        for(var i = 0; i < markerList.length; i++) {
+          if(markerList[i].pos == "end") { continue; }
+          var startMarker = markerList[i];
+          var endMarker = findEndOfComment(startMarker.id, markerList);
+
+          // Grab characters adjacent to endpoints of comment
+          var charAfterLast = codeMirror.getRange({line: endMarker.line, ch: endMarker.ch+1}, {line: endMarker.line, ch: endMarker.ch+2});
+          var charBeforeFirst = codeMirror.getRange({line: startMarker.line, ch: startMarker.ch-1}, {line: startMarker.line, ch: startMarker.ch});
+
+          // Add a space before and/or after the comment if there isn't one
+          if (charAfterLast != " ") {
+            codeMirror.replaceRange(" " + codeMirror.getRange({line: endMarker.line, ch: endMarker.ch+1}, {line: endMarker.line, ch: endMarker.ch+2}), {line: endMarker.line, ch: endMarker.ch+1}, {line: endMarker.line, ch: endMarker.ch+2});
+          }
+          if (charBeforeFirst != " ") {
+            codeMirror.replaceRange(codeMirror.getRange({line: startMarker.line, ch: startMarker.ch-1}, {line: startMarker.line, ch: startMarker.ch}) + " ", {line: startMarker.line, ch: startMarker.ch-1}, {line: startMarker.line, ch: startMarker.ch});
+            startMarker.ch++;
+            if (startMarker.line == endMarker.line) {
+              endMarker.ch++;
+            }
+          }
+
+          // Highlight the text and give it the id of the associated comment
+          codeMirror.markText({line: startMarker.line, ch: startMarker.ch-1}, {line: endMarker.line, ch: endMarker.ch+2}, {className: "comment " + startMarker.id});
+
+          // Make the start and end markers read only so that the user doesn't accidentally delete them
+          codeMirror.markText({line: endMarker.line, ch: endMarker.ch}, {line: endMarker.line, ch: endMarker.ch+2}, {readOnly: true});
+          codeMirror.markText({line: startMarker.line, ch: startMarker.ch-1}, {line: startMarker.line, ch: startMarker.ch+1}, {readOnly: true});
+        }
+
+        // Load comments themselves
+        var hash = "<%= (String)request.getAttribute("documentHash") %>"
```

Comment:
put this in a new function to modularize

Reply:
sounds good